loggly/loggly.go

```diff
@@ -76,59 +79,80 @@ func (l *Adapter) Stream(logstream chan *router.Message) {
 			ContainerImage:    m.Container.Config.Image,
 			ContainerHostname: m.Container.Config.Hostname,
 		}
-
-		l.SendMessage(msg)
+		a.enquequeMessage(msg)
 	}
 }
 
-// SendMessage handles creating and sending a request to Loggly. Any errors
-// that occur during that process are bubbled up to the caller
-func (l *Adapter) SendMessage(msg logglyMessage) {
-	js, err := json.Marshal(msg)
+// dispatcher sends bulks of bufferd logs every 2 seconds after a request to loggly
+func (a *Adapter) dispatcher() {
+	for {
+		time.Sleep(2 * time.Second)
+		if len(a.msgs) == 0 {
+			continue
+		}
+		a.dequeueMessages()
+	}
+}
+func (a *Adapter) enquequeMessage(msg logglyMessage) {
+	a.m.Lock()
+	a.msgs = append(a.msgs, msg)
+	a.m.Unlock()
+}
 
+// dequeueMessages marshals the log messages into json and cleas the buffer
+func (a *Adapter) dequeueMessages() {
+	a.m.Lock()
+	msgs := a.msgs
+	// clear log buffer
+	a.msgs = []logglyMessage{}
+	a.m.Unlock()
+
+	buf := &bytes.Buffer{}
+	for i, msg := range msgs {
+		msgJson, err := json.Marshal(msg)
+		if err != nil {
+			a.log.Println(err)
+			continue
+		}
+		_, err = buf.Write(msgJson)
+		if err != nil {
+			a.log.Println(err)
+			continue
+		}
+		// split the messages with a new line exept the last one
+		if i < len(msgs)-1 {
+			buf.WriteRune('\n')
+		}
+	}
+	err := a.sendLogs(buf)
 	if err != nil {
-		log.Println(err)
-		return
+		// if the send fialed requeue the log messages and send next time
+		log.Println("failed to rend logs to loggly, requeueing: ", err)
+		for _, msg := range msgs {
+			a.enquequeMessage(msg)
+		}
 	}
+}
 
-	url := fmt.Sprintf("%s%s/%s", logglyAddr, logglyEventEndpoint, l.token)
-	req, err := http.NewRequest("POST", url, bytes.NewBuffer(js))
+// sendLogs sends a bulk logs to loggly
+func (a *Adapter) sendLogs(logs io.Reader) error {
 
+	req, err := http.NewRequest("POST", a.url, logs)
 	if err != nil {
-		l.log.Println(err)
-		return
+		return err
 	}
 
-	if l.tags != "" {
-		req.Header.Add(logglyTagsHeader, l.tags)
+	if a.tags != "" {
+		req.Header.Add(logglyTagsHeader, a.tags)
 	}
 
-	go l.sendRequestToLoggly(req)
-}
-
-func (l *Adapter) sendRequestToLoggly(req *http.Request) {
-	resp, err := l.client.Do(req)
-
+	resp, err := a.client.Do(req)
 	if err != nil {
-		l.log.Println(
-			fmt.Errorf(
-				"error from client: %s",
-				err.Error(),
-			),
-		)
-		return
-	}
-	
-	defer resp.Body.Close()
-
-	if resp.StatusCode != http.StatusOK {
-		l.log.Println(
-			fmt.Errorf(
-				"received a non 200 status code when sending message to loggly: %s",
-				err.Error(),
-			),
-		)
+		a.log.Println("error from client: ", err)
+		return err
 	}
+	resp.Body.Close()
```

Comment:
This line won't get executed if there is an error and we'll like file descriptors